service/storage_service.cc

```diff
@@ -772,12 +772,6 @@ void storage_service::join_token_ring(int delay) {
         // start participating in the ring.
```

Comment:
Commenting on the commit message.

Are you sure that the replaced node is considered as a fat client (= gossip-only member)?

```
bool gossiper::is_gossip_only_member(inet_address endpoint) {
    auto es = get_endpoint_state_for_endpoint_ptr(endpoint);
    if (!es) {
        return false;
    }
    auto& ss = service::get_local_storage_service();
    return !is_dead_state(*es) && !ss.get_token_metadata().is_member(endpoint);
}
```

`es` will be engaged, `is_member()` will be false. But will `is_dead_state()` be false? As far as I know, you may be replacing a dead node, in which case `is_gossip_only_member()` will be false. So I don't think we can rely on the fat client timer and get rid of the 2x quarantine time.

Why did you have to get rid of the quarantine? It doesn't seem to be the source of discrepancy.

The source of discrepancy is that the replacing node evicts the replaced node from gossip, while other nodes do not. Do you know why does the replacing node evict immediately, why can't it keep it like other nodes?

Reply:
> Commenting on the commit message.
> 
> Are you sure that the replaced node is considered as a fat client (= gossip-only member)?

Yes.

> 
> ```
> bool gossiper::is_gossip_only_member(inet_address endpoint) {
>     auto es = get_endpoint_state_for_endpoint_ptr(endpoint);
>     if (!es) {
>         return false;
>     }
>     auto& ss = service::get_local_storage_service();
>     return !is_dead_state(*es) && !ss.get_token_metadata().is_member(endpoint);
> }
> ```
> 
> `es` will be engaged, `is_member()` will be false. But will `is_dead_state()` be false? As far as I know, you may be replacing a dead node, in which case `is_gossip_only_member()` will be false. So I don't think we can rely on the fat client timer and get rid of the 2x quarantine time.

For example:
                  
Before a node is replaced:
gossiper::is_gossip_only_member node=127.0.0.5, is_dead_state=0, is_member=1

After a node is replaced:
gossiper::is_gossip_only_member node=127.0.0.5, is_dead_state=0, is_member=0  
       
So, is_gossip_only_member returns true after the node is replaced. Note a node that is down is not in the dead_states, it is still in NORMAL status.

> Why did you have to get rid of the quarantine? It doesn't seem to be the source of discrepancy.

I did not get rid of the quarantine.  On storage_service::handle_state_normal, we sill call the storage_service::remove_endpoint which calls _gossiper.remove_endpoint quarantine the replaced node.


> The source of discrepancy is that the replacing node evicts the replaced node from gossip, while other nodes do not. 

Yes, "storage_service: Unify handling of replaced node removal from gossip" patch removes such discrepancy.

> Do you know why does the replacing node evict immediately, why can't it keep it like other nodes?

I do not see a real reason why the replacing node can't behave like the other nodes. I think it is this way because it is this way from the very beginning. The logic is removing the replaced node in the replacing node as soon as possible and preventing other nodes repopulating the replaced node info back. The replacing node is a little bit special because it is the one who actually replace the replaced node, so it does not need the gossip status handler to call the remove operation.

Reply:
> > Commenting on the commit message.
> > Are you sure that the replaced node is considered as a fat client (= gossip-only member)?
> 
> Yes.
> 
> > ```
> > bool gossiper::is_gossip_only_member(inet_address endpoint) {
> >     auto es = get_endpoint_state_for_endpoint_ptr(endpoint);
> >     if (!es) {
> >         return false;
> >     }
> >     auto& ss = service::get_local_storage_service();
> >     return !is_dead_state(*es) && !ss.get_token_metadata().is_member(endpoint);
> > }
> > ```
> > 
> > 
> > `es` will be engaged, `is_member()` will be false. But will `is_dead_state()` be false? As far as I know, you may be replacing a dead node, in which case `is_gossip_only_member()` will be false. So I don't think we can rely on the fat client timer and get rid of the 2x quarantine time.
> 
> For example:
> 
> Before a node is replaced:
> gossiper::is_gossip_only_member node=127.0.0.5, is_dead_state=0, is_member=1
> 
> After a node is replaced:
> gossiper::is_gossip_only_member node=127.0.0.5, is_dead_state=0, is_member=0
> 
> So, is_gossip_only_member returns true after the node is replaced. Note a node that is down is not in the dead_states, it is still in NORMAL status.

What if the replaced node was in one of the dead states when the replace started?

> 
> > Why did you have to get rid of the quarantine? It doesn't seem to be the source of discrepancy.
> 
> I did not get rid of the quarantine. 

I wasn't clear. I meant why did you have to revert the fix which makes the replacing node quarantine 2x longer. Now all nodes will quarantine for the same amount, which reverts the fix for CASSANDRA-8260.

It seems to me that we don't need to change this quarantine behavior in order to fix the gossip state discrepancy, which would make this change safer and easier to prove correct.

> 
> > The source of discrepancy is that the replacing node evicts the replaced node from gossip, while other nodes do not.
> 
> Yes, "storage_service: Unify handling of replaced node removal from gossip" patch removes such discrepancy.
> 
> > Do you know why does the replacing node evict immediately, why can't it keep it like other nodes?
> 
> I do not see a real reason why the replacing node can't behave like the other nodes. I think it is this way because it is this way from the very beginning. The logic is removing the replaced node in the replacing node as soon as possible and preventing other nodes repopulating the replaced node info back. The replacing node is a little bit special because it is the one who actually replace the replaced node, so it does not need the gossip status handler to call the remove operation.

I have some new thoughts about this.

The reason endpoint states don't converge is that they're ignored by gossip because they're quarantined. Why, then, not fix the problem by ignoring quarantined endpoints when calculating the hash? It seems it is the smallest fix which is also guaranteed to cover all such cases which may lead to discrepancy because we decided to quarantine an endpoint, not only replacing a node.

It also seems safer, since it doesn't change the core behavior of the gossiper (and changes there are hard to prove correct in general).

Reply:
> What if the replaced node was in one of the dead states when the replace started?

We can not replace a node in dead states. We can only replace a node that was in NORMAL status. 
```
  const std::vector<sstring> DEAD_STATES = {
      versioned_value::REMOVING_TOKEN,
      versioned_value::REMOVED_TOKEN,
      versioned_value::STATUS_LEFT,
      versioned_value::HIBERNATE
  };       
```

For REMOVING_TOKEN, the node is being removed by `nodetool removenode`, we should stick with that operation.

REMOVED_TOKEN and  STATUS_LEFT, the node is already out of the cluster.

For HIBERNATE, it is the replacing node it self.

Reply:
> I wasn't clear. I meant why did you have to revert the fix which makes the replacing node quarantine 2x longer. Now all nodes will quarantine for the same amount, which reverts the fix for CASSANDRA-8260.
> 
> It seems to me that we don't need to change this quarantine behavior in order to fix the gossip state discrepancy, which would make this change safer and easier to prove correct.

Note, the fix CASSANDRA-8260 was introduced to fix the bad decision to remove gossip state from the replacing node while keep the gossip state on other nodes. 

If we do not do that, the ugly fix is not needed at all.

Reply:
> I have some new thoughts about this.
> 
> The reason endpoint states don't converge is that they're ignored by gossip because they're quarantined. Why, then, not fix the problem by ignoring quarantined endpoints when calculating the hash? It seems it is the smallest fix which is also guaranteed to cover all such cases which may lead to discrepancy because we decided to quarantine an endpoint, not only replacing a node.
> 
> It also seems safer, since it doesn't change the core behavior of the gossiper (and changes there are hard to prove correct in general).

It does not change the core gossip behaviour it only changes the behaviour of how we handling the replacing node. 

I considered choosing different nodes to do the bash, but I gave up. 
The root cause of the discrepancy is due to removing gossip state on one node while keeping  on other nodes. This is very hard to follow and reason about, especially the 2x quarantined behaviour.  We should fix the root cause. With my fix, the behaviour is much much simpler.

Reply:
> > I wasn't clear. I meant why did you have to revert the fix which makes the replacing node quarantine 2x longer. Now all nodes will quarantine for the same amount, which reverts the fix for CASSANDRA-8260.
> > It seems to me that we don't need to change this quarantine behavior in order to fix the gossip state discrepancy, which would make this change safer and easier to prove correct.
> 
> Note, the fix CASSANDRA-8260 was introduced to fix the bad decision to remove gossip state from the replacing node while keep the gossip state on other nodes.
> 
> If we do not do that, the ugly fix is not needed at all.

I don't think that keeping the state on replacing node would be enough to fix CASSANDRA-8260 if you use the same quarantine period. If you read the sequence of events in https://issues.apache.org/jira/browse/CASSANDRA-8260?attachmentOrder=desc then notice that in step 4, when the quarantine on Z expires, the state would be removed. The bug does not depend on the state not being there before step 4.

Reply:
> > I have some new thoughts about this.
> > The reason endpoint states don't converge is that they're ignored by gossip because they're quarantined. Why, then, not fix the problem by ignoring quarantined endpoints when calculating the hash? It seems it is the smallest fix which is also guaranteed to cover all such cases which may lead to discrepancy because we decided to quarantine an endpoint, not only replacing a node.
> > It also seems safer, since it doesn't change the core behavior of the gossiper (and changes there are hard to prove correct in general).
> 
> It does not change the core gossip behaviour it only changes the behaviour of how we handling the replacing node.

I guess we mean different things behind "core behavior" :) What I meant is that this patch changes the algorithm of how replacing a node is handled. Changing this can introduce bugs like CASSANDRA-8260.


> 
> I considered choosing different nodes to do the bash, but I gave up.

Why?

> The root cause of the discrepancy is due to removing gossip state on one node while keeping on other nodes. This is very hard to follow and reason about, especially the 2x quarantined behaviour. We should fix the root cause. With my fix, the behaviour is much much simpler.

Reply:
> > Commenting on the commit message.
> > Are you sure that the replaced node is considered as a fat client (= gossip-only member)?
> 
> Yes.

I have another doubt about this. FatClient expiry is done like this:

```
 // check if this is a fat client. fat clients are removed automatically from
        // gossip after FatClientTimeout.  Do not remove dead states here.
        if (is_gossip_only_member(endpoint)
            && !_just_removed_endpoints.count(endpoint)
            && ((now - ep_state.get_update_timestamp()) > fat_client_timeout)) {
            logger.info("FatClient {} has been silent for {}ms, removing from gossip", endpoint, fat_client_timeout.count());
            remove_endpoint(endpoint); // will put it in _just_removed_endpoints to respect quarantine delay
            evict_from_membership(endpoint); // can get rid of the state immediately
        }
```

There is this check for `_just_removed_endpoints.count(endpoint)`, which means that quarantined endpoints are not considered for fat client expiry.

Reply:
Let me give you a full picture how this works:

Procedure is:
```
Remove node from token_metadata
Quarantine 60S
When Quarantine 60S is done, FatClient removal is performed
Removed from gossip endpoint_state_map
Quarantine 60S
```
The call chain is: 
```
storage_service::handle_state_normal -> token_metadata::remove_endpoint  ->  storage_service::remove_endpoint -> gossiper::remove_endpoint() -> gossiper::quarantine_endpoint()

fatclinet remove: gossiper::remove_endpoint() -> gossiper::evict_from_membership -> endpoint_state_map.erase(endpoint) - > gossiper::quarantine_endpoint()
```

Reply:
> > It does not change the core gossip behaviour it only changes the behaviour of how we handling the replacing node.
> 
> I guess we mean different things behind "core behavior" :) What I meant is that this patch changes the algorithm of how replacing a node is handled. Changing this can introduce bugs like CASSANDRA-8260.

Any change can cause bug. This patch simplify things a lot which is supposed to have less bugs.

> 
> > I considered choosing different nodes to do the bash, but I gave up.
> 
> Why?

Because I found a much better way to do so.

Reply:
> > Note, the fix CASSANDRA-8260 was introduced to fix the bad decision to remove gossip state from the replacing node while keep the gossip state on other nodes.
> > If we do not do that, the ugly fix is not needed at all.
> 
> I don't think that keeping the state on replacing node would be enough to fix CASSANDRA-8260 if you use the same quarantine period. If you read the sequence of events in https://issues.apache.org/jira/browse/CASSANDRA-8260?attachmentOrder=desc then notice that in step 4, when the quarantine on Z expires, the state would be removed. The bug does not depend on the state not being there before step 4.

The bug actually depends on state not being there before step 4. On the replacing node Z, the gossip state for Y is removed, that's why when X talks to Z, Z put Y in system table again. If Z still has Y or Z quarantines 120s, this won't happen. 

Also, think about another node W which similar to X. It uses the same method to remove Y.
X:  q60s  +  factclient_removal +  evict + q60s
W:  q60s  + factclient_removal  + evict + q60s
Z:   q60s  +  factclient_removal + evict + q60s

If we have problem in my new method to remove Z, we have the same problem with existing code on node W.

Reply:
> > > Note, the fix CASSANDRA-8260 was introduced to fix the bad decision to remove gossip state from the replacing node while keep the gossip state on other nodes.
> > > If we do not do that, the ugly fix is not needed at all.
> > 
> > 
> > I don't think that keeping the state on replacing node would be enough to fix CASSANDRA-8260 if you use the same quarantine period. If you read the sequence of events in https://issues.apache.org/jira/browse/CASSANDRA-8260?attachmentOrder=desc then notice that in step 4, when the quarantine on Z expires, the state would be removed. The bug does not depend on the state not being there before step 4.
> 
> The bug actually depends on state not being there before step 4. On the replacing node Z, the gossip state for Y is removed, that's why when X talks to Z, Z put Y in system table again. If Z still has Y or Z quarantines 120s, this won't happen.

Before step 4, Z is quarantining Y, so Z will not put info about Y from X before step 4, it will ignore it.

The problem happens in step 5 when X is talking to Z after the quarantine already expired on Z, that's where the problem happens.

Reply:
> > The bug actually depends on state not being there before step 4. On the replacing node Z, the gossip state for Y is removed, that's why when X talks to Z, Z put Y in system table again. If Z still has Y or Z quarantines 120s, this won't happen.
> 
> Before step 4, Z is quarantining Y, so Z will not put info about Y from X before step 4, it will ignore it.
> 
> The problem happens in step 5 when X is talking to Z after the quarantine already expired on Z, that's where the problem happens.

Of course, if Z is quarantining Y, Z will not update Y. I mean after the quarantine is over, if Z still has Y, it won't put Y in system table again, so having Y or not matters.

Anyway, could you think of any bug with my patch regarding the removal of Y?

Reply:
> > > The bug actually depends on state not being there before step 4. On the replacing node Z, the gossip state for Y is removed, that's why when X talks to Z, Z put Y in system table again. If Z still has Y or Z quarantines 120s, this won't happen.
> > 
> > 
> > Before step 4, Z is quarantining Y, so Z will not put info about Y from X before step 4, it will ignore it.
> > The problem happens in step 5 when X is talking to Z after the quarantine already expired on Z, that's where the problem happens.
> 
> Of course, if Z is quarantining Y, Z will not update Y. I mean after the quarantine is over, if Z still has Y, it won't put Y in system table again, so having Y or not matters.
> 

Wouldn't Z remove Y immediately when the quarantine is over? Because the fat client timer starts ticking when the quarantine starts, not when it's over, and because the fat client timeout is smaller than the quarantine period?

Reply:
> > > It does not change the core gossip behaviour it only changes the behaviour of how we handling the replacing node.
> > 
> > 
> > I guess we mean different things behind "core behavior" :) What I meant is that this patch changes the algorithm of how replacing a node is handled. Changing this can introduce bugs like CASSANDRA-8260.
> 
> Any change can cause bug. This patch simplify things a lot which is supposed to have less bugs.

In this case, we have a choice between two kinds of changes with different risk factors. That's not something to be ignored. I have been trying to prove this change to be correct for a while now and the amount of back and forth is an indication that this code is not trivial to prove correct, and hence high risk to change.

Reply:
> > Of course, if Z is quarantining Y, Z will not update Y. I mean after the quarantine is over, if Z still has Y, it won't put Y in system table again, so having Y or not matters.
> 
> Wouldn't Z remove Y immediately when the quarantine is over? Because the fat client timer starts ticking when the quarantine starts, not when it's over, and because the fat client timeout is smaller than the quarantine period?

The fatclient timeout starts once the node is not updating the timestamp. However, Z will not remove the node Y immediately after quarantine is over. It will do the fatclient removal only in the next gossip round: we remove the _just_removed_endpoints at the end of the gossip round. I think we should do the opposite: remove  _just_removed_endpoints before the fatclient removal. This could remove the race window between quarantine is over  and evict happens and re-quarantine.

Yes, fat client timeout is 0.5 * quarantine_delay = 30s, quarantine_delay = 2* ring_delay = 60s by default.

Reply:
> In this case, we have a choice between two kinds of changes with different risk factors. That's not something to be ignored. I have been trying to prove this change to be correct for a while now and the amount of back and forth is an indication that this code is not trivial to prove correct, and hence high risk to change.

Please note the fact, this patch changes the replacing node to behave as one of the existing node. If there was any issues, such issues exist even before this patch. It does not add new risks in my view.

Reply:
> > > Of course, if Z is quarantining Y, Z will not update Y. I mean after the quarantine is over, if Z still has Y, it won't put Y in system table again, so having Y or not matters.
> > 
> > 
> > Wouldn't Z remove Y immediately when the quarantine is over? Because the fat client timer starts ticking when the quarantine starts, not when it's over, and because the fat client timeout is smaller than the quarantine period?
> 
> The fatclient timeout starts once the node is not updating the timestamp. However, Z will not remove the node Y immediately after quarantine is over. It will do the fatclient removal only in the next gossip round: we remove the _just_removed_endpoints at the end of the gossip round. I think we should do the opposite: remove _just_removed_endpoints before the fatclient removal. This could remove the race window between quarantine is over and evict happens and re-quarantine.

Yes, I think that should do it.

So what we'd have is this on each node:

1. Z replaces X, keeps the state of X, quarantines
2. quarantine is over, state of X is removed, another quarantine starts
3. quarantine is over

I wonder though, why do we have to keep the state of X, and quarantine first with it, then without it?

Since Z replaces X, the state of Z should be enough. Other nodes when they see Z will know to remove X.

Another thing is, what will happen if Z is restarted during the quarantine? Will it lose the information about the quarantine?

Reply:
Perhaps the presence of Z in the endpoint states should be enough to reject any updates about X, quarantine should not be needed to prevent that.

Reply:
> Perhaps the presence of Z in the endpoint states should be enough to reject any updates about X, quarantine should not be needed to prevent that.

Z can have a different IP than X. From gossip point of view, it has no knowledge of that Z it is replacing so it should not create / update  X in endpoint_state_map.

Reply:
> Another thing is, what will happen if Z is restarted during the quarantine? Will it lose the information about the quarantine?

Yes, it will lose the quarantine info. But when Z boots again, it has two cases.  It learns X from other nodes, it does not learn X from other nodes if all has removed X already. In first case, When Z becomes NORMAL, it will figure out it replaces X. It starts the quarantine and evict process again. In second case, it does not matter because X is gone already.

Reply:
> > The fatclient timeout starts once the node is not updating the timestamp. However, Z will not remove the node Y immediately after quarantine is over. It will do the fatclient removal only in the next gossip round: we remove the _just_removed_endpoints at the end of the gossip round. I think we should do the opposite: remove _just_removed_endpoints before the fatclient removal. This could remove the race window between quarantine is over and evict happens and re-quarantine.
> 
> Yes, I think that should do it.
> 
> So what we'd have is this on each node:
> 
>     1. Z replaces X, keeps the state of X, quarantines
> 
>     2. quarantine is over, state of X is removed, another quarantine starts
> 
>     3. quarantine is over
> 
> 
> I wonder though, why do we have to keep the state of X, and quarantine first with it, then without it?
> 
> Since Z replaces X, the state of Z should be enough. Other nodes when they see Z will know to remove X.

Suppose each node removes the X  directly without the first quarantine. Say we have node Y, Z.  Z removes X directly. Y hasn't removed X yet and Y talks to Z and propagate X to Z. Now, Z will have X forever.